jbmc/src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -2335,105 +2335,78 @@ codet &java_bytecode_convert_methodt::do_exception_handling(
-      }
-      else
-        break;
-    }
-
-    // add the actual PUSH-CATCH instruction
-    if(!exception_ids.empty())
-    {
-      code_push_catcht catch_push;
-      INVARIANT(
-        exception_ids.size() == handler_labels.size(),
-        "Exception tags and handler labels should be 1-to-1");
-      code_push_catcht::exception_listt &exception_list =
-        catch_push.exception_list();
-      for(size_t i = 0; i < exception_ids.size(); ++i)
-      {
-        exception_list.push_back(
-          code_push_catcht::exception_list_entryt(
-            exception_ids[i], handler_labels[i]));
-      }
-
-      c = code_blockt({catch_push, c});
-    }
-    else
-    {
-      // advance
-      ++pos;
-    }
-
-    // reset
-    end_pc = 0;
-    exception_ids.clear();
-    handler_labels.clear();
+    if(exception.start_pc == cur_pc)
+      exceptions_by_end[exception.end_pc].push_back(exception);
+  }
+  for(const auto &exceptions : exceptions_by_end)
+  {
+    code_push_catcht catch_push;
+    code_push_catcht::exception_listt &exception_list =
+      catch_push.exception_list();
+    for(const java_bytecode_convert_methodt::methodt::exceptiont &exception
+      : exceptions.second)
+    {
+      exception_list.emplace_back(
+        exception.catch_type.get_identifier(),
+        // Record the exception handler in the CATCH-PUSH instruction by
+        // generating a label corresponding to the handler's pc
+        label(std::to_string(exception.handler_pc)));
+    }
+    // Prepend the CATCH-PUSH instruction
+    code = code_blockt({ catch_push, code });
   }
 
-  // next add the CATCH-POP instructions
-  size_t start_pc = 0;
-  // as the first try block may have start_pc 0, we
-  // must track it separately
-  bool first_handler = false;
-  // check if this is the end of a try block
+  // Next add the CATCH-POP instructions
+  // exception_row.end_pc is exclusive so append CATCH-POP instructions if this
+  // is the instruction before it.
+  // Dangerously, we assume that the lexically next bit of code after the try
+  // block (the catch block) is:
+  // a) the next code in the bytecode and therefore its start address matches
+  // the exclusive end position of the try block (probably a valid assumption
+  // for non-obfuscated bytecode generated by most compilers)
+  // b) the successor of a previous instruction and therefore in the working
+  // set (not a valid assumption when there aren't any throwing statements in
+  // the try block)
+  // We are therefore going to find all exception handlers that end at our
+  // earliest successor
+  if(working_set.empty())
+  {
+    // There was no throw in the try block
+    // TODO: This will result in non-balanced PUSH and POPs
+    return code;
+  }
+  const method_offsett &working_set_elt = *working_set.begin();
+  if(cur_pc >= working_set_elt)
+  {
+    // This just fails if we have a back-edge here
+    // TODO: This will result in non-balanced PUSH and POPs
+    return code;
```

Comment:
:question: Ditto here. Also might be a good idea to print some warning in this eventuallity?